File: register_obsidian_vault.py
import json
import os
import sys
import time
import uuid
import traceback

# Debug log file
DEBUG_LOG = os.path.join(os.getenv('TEMP', '.'), 'obsidian_vault_register_debug.log')

def log_debug(message):
    """Write debug message to log file."""
    try:
        with open(DEBUG_LOG, 'a', encoding='utf-8') as f:
            timestamp = time.strftime('%Y-%m-%d %H:%M:%S')
            f.write(f"[{timestamp}] {message}\n")
    except:
        pass  # Fail silently if we can't write debug log

def get_appdata_path():
    """Returns the path to the user's AppData\Roaming directory."""
    return os.getenv('APPDATA')

def get_obsidian_json_path():
    """Returns the full path to the obsidian.json file."""
    appdata_path = get_appdata_path()
    if not appdata_path:
        raise EnvironmentError("APPDATA environment variable not found.")
    return os.path.join(appdata_path, 'Obsidian', 'obsidian.json')

def register_vault(vault_path):
    """Adds a new vault to obsidian.json if it doesn't already exist."""
    log_debug(f"=== Starting registration for: {vault_path}")

    obsidian_json_path = get_obsidian_json_path()
    log_debug(f"Obsidian JSON path: {obsidian_json_path}")

    # Ensure the Obsidian directory exists
    os.makedirs(os.path.dirname(obsidian_json_path), exist_ok=True)
    log_debug(f"Ensured directory exists: {os.path.dirname(obsidian_json_path)}")

    vault_path_normalized = os.path.abspath(vault_path)
    log_debug(f"Normalized vault path: {vault_path_normalized}")

    data = {"vaults": {}, "openSchemes": {"app": True}}
    
    if os.path.exists(obsidian_json_path):
        log_debug(f"Obsidian JSON exists, reading...")
        try:
            with open(obsidian_json_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            log_debug(f"Successfully read obsidian.json")
        except json.JSONDecodeError as e:
            log_debug(f"JSON decode error: {e}")
            print(f"Warning: Could not decode existing obsidian.json. Creating new one.")
            # If corrupted, start with an empty structure
            data = {"vaults": {}, "openSchemes": {"app": True}}
    else:
        log_debug(f"Obsidian JSON does not exist, will create new one")

    vaults = data.get("vaults", {})
    log_debug(f"Existing vaults count: {len(vaults)}")

    # Check if the vault already exists by path
    for vault_id, vault_info in vaults.items():
        existing_path = os.path.abspath(vault_info.get("path", ""))
        log_debug(f"Checking vault {vault_id}: {existing_path}")
        if existing_path == vault_path_normalized:
            log_debug(f"Found existing vault with ID: {vault_id}")
            print(f"VAULT_ID:{vault_id}")
            return vault_id
            
    # If not found, add it
    log_debug("Vault not found, creating new entry")
    new_vault_id = uuid.uuid4().hex
    # Ensure the ID is exactly 16 hex chars as Obsidian uses
    while len(new_vault_id) < 16:
        new_vault_id += '0' # Pad with '0' if UUID.hex is less than 16 (unlikely but safe)
    new_vault_id = new_vault_id[:16] # Take first 16 chars
    log_debug(f"Generated vault ID: {new_vault_id}")

    vaults[new_vault_id] = {
        "path": vault_path_normalized,
        "ts": int(time.time() * 1000)
    }
    log_debug(f"Added vault to vaults dictionary")

    data["vaults"] = vaults

    try:
        with open(obsidian_json_path, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2)
        log_debug(f"Successfully wrote to obsidian.json")
    except Exception as e:
        log_debug(f"ERROR writing to obsidian.json: {e}")
        raise

    print(f"VAULT_ID:{new_vault_id}")
    log_debug(f"Registration complete, vault ID: {new_vault_id}")
    return new_vault_id

if __name__ == "__main__":
    log_debug(f"====== Script started ======")
    log_debug(f"Arguments: {sys.argv}")
    log_debug(f"Debug log location: {DEBUG_LOG}")

    if len(sys.argv) < 2:
        log_debug("ERROR: No vault path provided")
        print("Usage: python register_obsidian_vault.py <vault_path>")
        sys.exit(1)

    vault_path = sys.argv[1]
    log_debug(f"Vault path argument: {vault_path}")

    # Ensure vault path is a directory and exists
    if not os.path.isdir(vault_path):
        log_debug(f"ERROR: Vault path is not a valid directory: {vault_path}")
        print(f"Error: Vault path '{vault_path}' is not a valid directory.")
        sys.exit(1)

    try:
        register_vault(vault_path)
        log_debug("====== Script completed successfully ======")
    except Exception as e:
        log_debug(f"EXCEPTION: {e}")
        log_debug(f"Traceback: {traceback.format_exc()}")
        print(f"An error occurred: {e}")
        sys.exit(1)

File: register_obsidian_vault.js
// JScript for registering Obsidian vaults
// Uses built-in Windows Script Host (cscript.exe)

var fso = new ActiveXObject("Scripting.FileSystemObject");
var shell = new ActiveXObject("WScript.Shell");

// Get command line arguments
var args = WScript.Arguments;
if (args.Length < 1) {
    WScript.Echo("Usage: cscript register_obsidian_vault.js <vault_path>");
    WScript.Quit(1);
}

var vaultPath = args(0);

// Debug log file
var tempFolder = shell.ExpandEnvironmentStrings("%TEMP%");
var debugLog = tempFolder + "\\obsidian_vault_register_debug.log";

function writeDebugLog(message) {
    try {
        var logFile = fso.OpenTextFile(debugLog, 8, true); // 8 = ForAppending
        var timestamp = new Date().toISOString().replace('T', ' ').substr(0, 19);
        logFile.WriteLine("[" + timestamp + "] " + message);
        logFile.Close();
    } catch (e) {
        // Fail silently
    }
}

// JSON parser for JScript (which doesn't have native JSON support)
function parseJSON(jsonString) {
    // Use eval with safety wrapper
    return eval('(' + jsonString + ')');
}

function stringifyJSON(obj) {
    // Simple JSON stringifier for JScript
    if (obj === null) return 'null';
    if (obj === undefined) return undefined;

    var type = typeof obj;
    if (type === 'number' || type === 'boolean') return '' + obj;
    if (type === 'string') return '"' + obj.replace(/\\/g, '\\\\').replace(/"/g, '\\"') + '"';

    if (obj instanceof Array || obj.length !== undefined) {
        var arr = [];
        for (var i = 0; i < obj.length; i++) {
            arr.push(stringifyJSON(obj[i]));
        }
        return '[' + arr.join(',') + ']';
    }

    if (type === 'object') {
        var properties = [];
        for (var key in obj) {
            if (obj.hasOwnProperty(key)) {
                properties.push(stringifyJSON(key) + ':' + stringifyJSON(obj[key]));
            }
        }
        return '{' + properties.join(',') + '}';
    }
}

function getObsidianJsonPath() {
    var appData = shell.ExpandEnvironmentStrings("%APPDATA%");
    return appData + "\\Obsidian\\obsidian.json";
}

function generateVaultId() {
    // Generate 16 hex characters
    var chars = "0123456789abcdef";
    var id = "";
    for (var i = 0; i < 16; i++) {
        id += chars.charAt(Math.floor(Math.random() * 16));
    }
    return id;
}

function registerVault(vaultPath) {
    writeDebugLog("=== Starting registration for: " + vaultPath);

    var obsidianJsonPath = getObsidianJsonPath();
    writeDebugLog("Obsidian JSON path: " + obsidianJsonPath);

    // Ensure directory exists
    var obsidianDir = fso.GetParentFolderName(obsidianJsonPath);
    if (!fso.FolderExists(obsidianDir)) {
        fso.CreateFolder(obsidianDir);
        writeDebugLog("Created directory: " + obsidianDir);
    }

    // Normalize path
    var vaultPathNormalized = fso.GetAbsolutePathName(vaultPath);
    writeDebugLog("Normalized vault path: " + vaultPathNormalized);

    // Read existing JSON
    var data;
    var existingVaultsCount = 0;

    if (fso.FileExists(obsidianJsonPath)) {
        writeDebugLog("Obsidian JSON exists, reading...");

        // Create backup
        var backupPath = obsidianJsonPath + ".backup";
        fso.CopyFile(obsidianJsonPath, backupPath, true);
        writeDebugLog("Created backup at: " + backupPath);

        try {
            // Use FSO to read as ASCII (matches how we write)
            var file = fso.OpenTextFile(obsidianJsonPath, 1, false, 0); // ForReading, ASCII
            var jsonText = file.ReadAll();
            file.Close();

            data = parseJSON(jsonText);
            writeDebugLog("Successfully parsed obsidian.json");

            // Count existing vaults
            if (data.vaults) {
                for (var key in data.vaults) {
                    if (data.vaults.hasOwnProperty(key)) {
                        existingVaultsCount++;
                    }
                }
            }
            writeDebugLog("Existing vaults count: " + existingVaultsCount);

        } catch (e) {
            writeDebugLog("ERROR reading JSON: " + e.message);
            WScript.Echo("ERROR: Could not read existing obsidian.json");
            WScript.Quit(1);
        }
    } else {
        writeDebugLog("Obsidian JSON does not exist, will create new one");
        data = {
            vaults: {},
            openSchemes: { app: true }
        };
    }

    // Ensure vaults object exists
    if (!data.vaults) {
        data.vaults = {};
    }

    // Check if vault already exists
    for (var vaultId in data.vaults) {
        if (data.vaults.hasOwnProperty(vaultId)) {
            var vaultInfo = data.vaults[vaultId];
            var existingPath = vaultInfo.path;

            writeDebugLog("Checking vault " + vaultId + ": " + existingPath);

            // Normalize for comparison
            try {
                existingPath = fso.GetAbsolutePathName(existingPath);
            } catch (e) {
                // Path might not exist anymore
            }

            if (existingPath.toLowerCase() === vaultPathNormalized.toLowerCase()) {
                writeDebugLog("Found existing vault with ID: " + vaultId);
                WScript.Echo("VAULT_ID:" + vaultId + ":EXISTING");
                WScript.Quit(0);
            }
        }
    }

    // Generate new vault ID
    writeDebugLog("Vault not found, creating new entry");
    var newVaultId = generateVaultId();
    writeDebugLog("Generated vault ID: " + newVaultId);

    // Add new vault
    data.vaults[newVaultId] = {
        path: vaultPathNormalized,
        ts: new Date().getTime()
    };

    // Count vaults after adding
    var newVaultsCount = 0;
    for (var key in data.vaults) {
        if (data.vaults.hasOwnProperty(key)) {
            newVaultsCount++;
        }
    }

    writeDebugLog("Vaults after adding new: " + newVaultsCount + " (was: " + existingVaultsCount + ")");

    // Safety check
    if (newVaultsCount < existingVaultsCount) {
        writeDebugLog("ERROR: Vault count decreased! Aborting write.");
        WScript.Echo("ERROR: Would lose existing vaults. Aborting.");
        WScript.Quit(1);
    }

    // Write JSON (compact format, UTF-8 without BOM)
    try {
        var jsonOutput = stringifyJSON(data);

        // Use FSO to write as ASCII (no BOM added)
        // Note: This works because obsidian.json only contains ASCII characters
        // (backslashes are escaped as \\ in JSON)
        var textFile = fso.CreateTextFile(obsidianJsonPath, true, false); // overwrite=true, unicode=false
        textFile.Write(jsonOutput);
        textFile.Close();

        writeDebugLog("Successfully wrote to obsidian.json (ASCII, no BOM)");
        writeDebugLog("Final vault count: " + newVaultsCount);

    } catch (e) {
        writeDebugLog("ERROR writing to obsidian.json: " + e.message);
        WScript.Echo("ERROR: Failed to write obsidian.json");
        WScript.Quit(1);
    }

    WScript.Echo("VAULT_ID:" + newVaultId + ":NEW");
    writeDebugLog("Registration complete, vault ID: " + newVaultId);
}

// Main execution
writeDebugLog("====== Script started ======");
writeDebugLog("Arguments: " + vaultPath);
writeDebugLog("Debug log location: " + debugLog);

// Check if path exists and is a directory
if (!fso.FolderExists(vaultPath)) {
    writeDebugLog("ERROR: Vault path is not a valid directory: " + vaultPath);
    WScript.Echo("Error: Vault path '" + vaultPath + "' is not a valid directory.");
    WScript.Quit(1);
}

try {
    registerVault(vaultPath);
    writeDebugLog("====== Script completed successfully ======");
    WScript.Quit(0);
} catch (e) {
    writeDebugLog("EXCEPTION: " + e.message);
    WScript.Echo("An error occurred: " + e.message);
    WScript.Quit(1);
}
